### PR TITLE
isnan operator

### LINKED ISSUE #1037: Implement ONNX Operator IsNaN
Needed for some recent model support requests. The original models are from TensorFlow. The tf2onnx tool indicates that the TF Operator "isNAN" has paths that call ONNX Operator isNAN

### PR DESCRIPTION
**Pull request by CharlieL7**
Implements the IsNaN operator, ref, gpu, and onnx parser.

Resolves #1037

### CODE REVIEW

File: src/include/migraphx/op/isnan.hpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#include <migraphx/op/unary.hpp>
+#include <migraphx/config.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace op {
+
+struct isnan : unary<isnan>
+{
+    std::string point_function() const { return "isnan"; }
```
Comment (pfultz2) [inline src/include/migraphx/op/isnan.hpp:13]: This shouldn't be needed as `unary` will use operator name for the `point_function`.

File: src/include/migraphx/op/isnan.hpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    auto apply() const
+    {
+        return [](auto x) { return std::isnan(x); };
+    }
+
+    std::string name() const { return "isnan"; }
+
+    shape compute_shape(std::vector<shape> inputs) const
```
Comment (pfultz2) [inline src/include/migraphx/op/isnan.hpp:21]: This is already handled by the `unary<isnan>` base class.

File: test/verify/test_isnan.cpp [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#include <limits>
+#include "verify_program.hpp"
+#include <migraphx/program.hpp>
+#include <migraphx/generate.hpp>
+#include <migraphx/make_op.hpp>
+
+struct test_abs : verify_program<test_abs>
```
Comment (pfultz2) [inline test/verify/test_isnan.cpp:7]: This needs be name `test_isnan`.

File: test/verify/test_isnan.cpp [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+{
+    migraphx::program create_program() const
+    {
+        migraphx::program p;
+        auto* mm = p.get_main_module();
+        auto x   = mm->add_parameter("x", migraphx::shape{migraphx::shape::float_type, {2, 3}});
```
Comment (pfultz2) [inline test/verify/test_isnan.cpp:13]: I think you can drop the second dimension and just do `migraphx::shape{migraphx::shape::float_type, {2}}`

File: test/verify/test_isnan.cpp [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        std::vector<float> data0{3, std::numeric_limits<float>::quiet_NaN()};
+        migraphx::shape s0{migraphx::shape::float_type, {1, 3}};
```
Comment (pfultz2) [inline test/verify/test_isnan.cpp:15]: Same here you can do `{migraphx::shape::float_type, {1}}`.

File: test/verify/test_isnan.cpp [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        auto l0 = mm->add_literal(migraphx::literal{s0, data0});
```
Comment (pfultz2) [inline test/verify/test_isnan.cpp:16]: I think you can just do `mm->add_literal(std::numeric_limits<float>::quiet_NaN())`.

File: test/verify/test_isnan.cpp [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+        x       = mm->add_instruction(migraphx::make_op("concat", {{"axis", 0}}), x, l0);
+        mm->add_instruction(migraphx::make_op("isnan"), x);
+        return p;
+    }
+};
```
Comment (pfultz2) [inline test/verify/test_isnan.cpp:21]: There should also be a test using half as well.

File: test/ref_ops_test.cpp [hunk 8]
```diff
@@ -1946,6 +1947,45 @@ TEST_CASE(if_pl_test)
...
+TEST_CASE(isnan_test)
+{
+    // float test
+    {
+        migraphx::program p;
+        auto* mm = p.get_main_module();
+        migraphx::shape s{migraphx::shape::float_type, {2, 3}};
+        auto nan_val             = std::numeric_limits<float>::quiet_NaN();
+        std::vector<float> data0 = {1.2, 5.2, nan_val, nan_val, 0., 100.};
+        auto l1                  = mm->add_literal(migraphx::literal{s, data0});
```
Comment (pfultz2) [inline test/ref_ops_test.cpp:1959]: I think you can do `mm->add_literal(migraphx::literal{s, {1.2, 5.2, nan_val, nan_val, 0., 100.}})`.

File: src/include/migraphx/op/isnan.hpp (hunk 2 above)
Comment (turneram) [reply src/include/migraphx/op/isnan.hpp:21]: The compute_shape method in unary.hpp was returning a shape with the same type as the input, but ONNX specifies that the output is bool.

File: src/CMakeLists.txt [hunk 9]
```diff
@@ -114,6 +114,7 @@ register_migraphx_ops(
     identity
     if_op
     im2col
+	isnan
```
Comment (umangyadav) [inline src/CMakeLists.txt:117]: Can you remove extra indentation ?

File: test/verify/test_isnan.cpp (hunk 3 above)
Comment (umangyadav) [inline test/verify/test_isnan.cpp:7]: You can also add a test for one of the nonstandard shapes e.g. broadcast, slice, or transpose.

File: src/include/migraphx/op/isnan.hpp (hunk 2 above)
Comment (pfultz2) [reply src/include/migraphx/op/isnan.hpp:21]: Then call `unary<isnan>` first and then change the type , but we need to add a `with_type` method to the shape class.

File: test/ref_ops_test.cpp (hunk 8 above)
Comment (CharlieL7) [reply test/ref_ops_test.cpp:1959]: There looks to be a problem with deducing the vector template type with the nan_val.

File: src/include/migraphx/op/isnan.hpp (hunk 2 above)
Comment (pfultz2) [reply src/include/migraphx/op/isnan.hpp:21]: Well I added `with_type` in #1102.

Comment (CharlieL7) [reply src/include/migraphx/op/isnan.hpp:20]: I pulled the branch, looks to work by hiding and calling the base class.

File: test/ref_ops_test.cpp [hunk 10]
```diff
@@ -1946,6 +1947,65 @@ TEST_CASE(if_pl_test)
...
+        auto result = p.eval({}).back();
+        std::vector<float> results_vector;
+        result.visit([&](auto output) { results_vector.assign(output.begin(), output.end()); });
+        std::vector<float> correct = {0, 0, 1, 1, 0, 0};
+        EXPECT(migraphx::verify_range(results_vector, correct));
+    }
+
+    // broadcast test
+    {
+        migraphx::program p;
```
Comment (umangyadav) [inline test/ref_ops_test.cpp:1990]: Nonstandard shape tests are in `ref_ops_nonstd_shape_test.cpp`. You can move this test there instead.

**Review by turneram — APPROVED** [review]

File: src/include/migraphx/shape.hpp [hunk 11]
```diff
@@ -131,6 +131,8 @@ struct shape
     shape with_lens(type_t t, const std::vector<std::size_t>& l) const;
     shape with_lens(const std::vector<std::size_t>& l) const;
 
+    shape with_type(type_t t) const;
```
Comment (umangyadav) [inline src/include/migraphx/shape.hpp:134]: Can you rebase with develop and push changes again ? Paul's changes are showing up here. Otherwise, LGTM.

**Review by umangyadav — APPROVED** [review]

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**